hibernate-reactive-core/src/main/java/org/hibernate/reactive/event/impl/AbstractReactiveSaveEventListener.java

```diff
@@ -162,34 +163,44 @@ protected CompletionStage<Void> reactivePerformSave(
 			LOG.tracev( "Saving {0}", infoString( persister, id, source.getFactory() ) );
 		}
 
-		final EntityKey key;
+		return generateEntityKey( entity, id, persister, useIdentityColumn, source ).thenCompose( key -> {
+			if ( !useIdentityColumn ) {
+				persister.setIdentifier( entity, id, source );
+			}
+			return reactivePerformSaveOrReplicate(
+				entity,
+				key,
+				persister,
+				useIdentityColumn,
+				context,
+				source,
+				requiresImmediateIdAccess );
+		} );
+	}
```

Comment:
IMO, this flow is just slightly convoluted. I would prefer to use something like:

```java
CompletionStage<EntityKey> key = !useIdentityColumn
        ? generateEntityKey( entity, id, persister, useIdentityColumn, source ).thenAccept( key -> persister.setIdentifier( entity, id, source ) )
        : CompletionStages.nullFuture();
return key.thenCompose( key -> reactivePerformSaveOrReplicate(
 				entity,
 				key,
 				persister,
 				useIdentityColumn,
 				context,
 				source,
 				requiresImmediateIdAccess ) );

```

That's simpler and clearer IMO.

Reply:
(And then the logic in `generateEntityKey()` can also be simplified.

Reply:
I ran with your version and ended up with the following that works too with basically the same #lines of code. I prefer the first version (after removing the redundant `useIdentityColumn` parameter in `generateEntityKey(...)`)

... and I could live with either.
```
	protected CompletionStage<Void> reactivePerformSave(
			Object entity,
			Serializable id,
			EntityPersister persister,
			boolean useIdentityColumn,
			C context,
			EventSource source,
			boolean requiresImmediateIdAccess) {

		if ( LOG.isTraceEnabled() ) {
			LOG.tracev( "Saving {0}", infoString( persister, id, source.getFactory() ) );
		}

		CompletionStage<EntityKey> key = useIdentityColumn
				? CompletionStages.nullFuture()
				: generateEntityKey( id, persister, source ).thenCompose( generatedKey -> {
			persister.setIdentifier( entity, id, source );
			return CompletionStages.completedFuture( generatedKey );
		} );

		return key.thenCompose( v -> {
			try {
				return reactivePerformSaveOrReplicate(
						entity,
						key.toCompletableFuture().get(),
						persister,
						useIdentityColumn,
						context,
						source,
						requiresImmediateIdAccess
				);
			}
			catch (Exception e) {
				return failedFuture( e );
			}
		} );
	}

	private CompletionStage<EntityKey> generateEntityKey(Serializable id, EntityPersister persister, EventSource source) {
		final EntityKey key = source.generateEntityKey( id, persister );
		final PersistenceContext persistenceContext = source.getPersistenceContextInternal();
		Object old = persistenceContext.getEntity( key );
		if ( old != null ) {
			if ( persistenceContext.getEntry( old ).getStatus() == Status.DELETED ) {
				return source.unwrap( ReactiveSession.class ).reactiveForceFlush(
						persistenceContext.getEntry( old ) ).thenApply( v -> key );
			}
			else {
				return failedFuture( new NonUniqueObjectException( id, persister.getEntityName() ) );
			}
		}

		return CompletionStages.completedFuture( key );
	}
```

Reply:
I think theres two things wrong there:

- you use `thenCompose()` instead of `thenAccept()` which is redundant, and bad style, and
- I'm not sure what the point of the try/catch is ... the method you're calling wouldn't throw exceptions synchronously so there's nothing to catch.

Reply:
(So if you eliminate the try/catch, and use `thenAccept()` then both those lambdas can be turned into one-line lambdas.)

Reply:
You can use `thenApply`:
```
CompletionStage<EntityKey> key = useIdentityColumn
				? CompletionStages.nullFuture()
				: generateEntityKey( id, persister, source ).thenApply generatedKey -> {
			persister.setIdentifier( entity, id, source );
			return generatedKey;
		} );
```

Reply:
You don't need the try-catch:
```
return key.thenCompose( entityKey -> reactivePerformSaveOrReplicate(
 		entity,
 		entityKey,
 		persister,
 		useIdentityColumn,
 		context,
 		source,
 		requiresImmediateIdAccess 
) );
```

Reply:
fixed